File: java/advent2022/src/main/java/org/mersic/Day02.java
package org.mersic;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

public class Day02 {
    private static final Map<String, Integer> SCORE_P1 = new HashMap<>();
    private static final Map<String, Integer> SCORE_P2 = new HashMap<>();
    static {
        //ROCK
        SCORE_P1.put("A X", 1+3);
        SCORE_P1.put("A Y", 2+6);
        SCORE_P1.put("A Z", 3+0);

        //PAPER
        SCORE_P1.put("B X", 1+0);
        SCORE_P1.put("B Y", 2+3);
        SCORE_P1.put("B Z", 3+6);

        //SCISSORS
        SCORE_P1.put("C X", 1+6);
        SCORE_P1.put("C Y", 2+0);
        SCORE_P1.put("C Z", 3+3);

        //ROCK 
        SCORE_P2.put("A X", 3+0);
        SCORE_P2.put("A Y", 1+3);
        SCORE_P2.put("A Z", 2+6);

        //PAPER
        SCORE_P2.put("B X", 1+0);
        SCORE_P2.put("B Y", 2+3);
        SCORE_P2.put("B Z", 3+6);

        //SCISSORS
        SCORE_P2.put("C X", 2+0);
        SCORE_P2.put("C Y", 3+3);
        SCORE_P2.put("C Z", 1+6);
    }
    
    public static void main(String[] args) throws Exception {
        record Scores(int s1, int s2) {}
        Scores s = Files.readAllLines(Path.of(Day02.class.getClassLoader().getResource("day.02.input").toURI())).stream()
                .map(x -> new Scores(SCORE_P1.get(x), SCORE_P2.get(x)))
                .reduce((x,y)->new Scores(x.s1 + y.s1, x.s2 + y.s2)).get();

        System.out.println("day 2 part 1: " + s.s1);
        System.out.println("day 2 part 2: " + s.s2);
    }
}

File: kotlin/advent2022/src/main/kotlin/Day02.kt
import java.nio.file.Files
import java.nio.file.Path

private val SCORE_P1 = mapOf(
    "A X" to 1+3, "A Y" to 2+6, "A Z" to 3+0,
    "B X" to 1+0, "B Y" to 2+3, "B Z" to 3+6,
    "C X" to 1+6, "C Y" to 2+0, "C Z" to 3+3,
)

private val SCORE_P2 = mapOf(
    "A X" to 3+0, "A Y" to 1+3, "A Z" to 2+6,
    "B X" to 1+0, "B Y" to 2+3, "B Z" to 3+6,
    "C X" to 2+0, "C Y" to 3+3, "C Z" to 1+6,
)


fun main() {
    val scores = Files.readAllLines(Path.of(Stub().javaClass.getClassLoader().getResource("day.02.input").toURI())).stream()
        .map { x -> Pair(SCORE_P1[x], SCORE_P2[x]) }
        .reduce(Pair(0, 0)) { acc, x -> Pair(acc.first!! + x.first!!, acc.second!! + x.second!!) }

    println("day 2 part 1: " + scores.first)
    println("day 2 part 2: " + scores.second)
}